### PR TITLE
pgwire: pad char values with spaces if length is less than column width

### LINKED ISSUE #49639: sql: column padding not preserved like in postgres
tldr: postgres preserves the right-padding of padded columns in some cases, CockroachDB does not.

Start with the following SQL

```
> CREATE TABLE t(x CHAR(8));
> INSERT INTO t(x) VALUES ('hello');
```

The run `SELECT x FROM t` in crdb and pg.

PostgreSQL:

```
{"Type":"DataRow","Values":[{"text":"hello   "}]}
```

CockroachDB:

```
{"Type":"DataRow","Values":[{"text":"hello"}]}
```

(I found this discrepancy when running the `row_description` pgtest in both crdb and pg)

It's hard to see in a SQL shell because the whitespace does not show on the screen, but still.

Interestingly, the following query behaves differently:

```
SELECT ':' || x || ':' FROM t
```

This returns both a short `:hello:` in both pg and crdb.

### PR DESCRIPTION
**Pull request by arulajmani**
Previously, when returning row values of a column which has type char,
the value wouldn't be padded with spaces on the right if its length
was less than the column width.

This is in contrast to Postgres, which pads values with spaces on the
right to match the column width. This change fixes this by checking
the Oid and width of the value being written, and padding it
appropriately if required.

Fixes #49639

Release note (bug fix): Previously, when streaming values from a column
declared of type char(n), the length of the value could be <= n. After
this change, all values streamed have length exactly n by padding
spaces to the right if necessary.

### CODE REVIEW

File: pkg/sql/pgwire/command_result.go [hunk 1]
```diff
@@ -75,9 +76,9 @@ type commandResult struct {
 	// to have an entry for every column.
 	formatCodes []pgwirebase.FormatCode
 
-	// oids is a map from result column index to its Oid, similar to formatCodes
-	// (except oids must always be set).
-	oids []oid.Oid
+	// types is a map from result column index to its type T, similar to formatCodes
+	// (except types must always be set).
+	types []types.T
```
Comment (rohany) [inline pkg/sql/pgwire/command_result.go:81]: You should use `*types.T` instead, theres no real reason to copy it.

File: pkg/sql/pgwire/testdata/pgtest/row_description [hunk 2]
```diff
@@ -116,7 +116,7 @@ until
 ReadyForQuery
 ----
 {"Type":"RowDescription","Fields":[{"Name":"b","TableOID":55,"TableAttributeNumber":2,"DataTypeOID":1042,"DataTypeSize":-1,"TypeModifier":12,"Format":0}]}
-{"Type":"DataRow","Values":[{"text":"hello"}]}
+{"Type":"DataRow","Values":[{"text":"hello   "}]}
```
Comment (rohany) [inline pkg/sql/pgwire/testdata/pgtest/row_description:119]: Are these meant to be changed? The comment for the bug is below at line 172

File: pkg/sql/pgwire/types.go [hunk 3]
```diff
@@ -116,7 +120,11 @@ func (b *writeBuffer) writeTextDatum(
 		b.writeLengthPrefixedString(v.IPAddr.String())
 
 	case *tree.DString:
-		b.writeLengthPrefixedString(string(*v))
+		sv := string(*v)
+		if T.Oid() == oid.T_bpchar {
```
Comment (rohany) [inline pkg/sql/pgwire/types.go:124]: Just checking, bpchar is the opposite of varchar right (using fixed width)? If so, this needs a comment, and so does the formatting directive here (idk what this does)

File: pkg/sql/pgwire/types.go [hunk 4]
```diff
@@ -398,7 +406,11 @@ func (b *writeBuffer) writeBinaryDatum(
 		b.writeLengthPrefixedString(v.LogicalRep)
 
 	case *tree.DString:
-		b.writeLengthPrefixedString(string(*v))
+		sv := string(*v)
+		//if T.Oid() == oid.T_bpchar {
```
Comment (rohany) [inline pkg/sql/pgwire/types.go:410]: commented code?

**Review by rohany — CHANGES_REQUESTED** [review]

**Comment by rohany** [comment]
Huh, why is blathers marking this with o-community

**Review by rafiss — CHANGES_REQUESTED** [review]
thanks! mostly just a few nits, and a question about testing writeBinaryDatum

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49886#-:-M90I0XZ5ZJtC0S-dTJu:b-3tx3ky)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @arulajmani, @rafiss, and @rohany)

___
*[pkg/sql/pgwire/types.go, line 72 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/49886#-M90HU_t78nL-8OVI6CE:-M90HU_t78nL-8OVI6CF:b-co54sx) ([raw file](https://github.com/cockroachdb/cockroach/blob/52cf6aa18e52b5441da64b2430a99d1ec4c1ceff/pkg/sql/pgwire/types.go#L72)):*
> ```
> T
> ```
nit: the parameter name should be a lower-case `t`, though I'd even be more in favor of using the name `type`.
___
*[pkg/sql/pgwire/types.go, line 200 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/49886#-M90GoSLBa0HauEPFn1_:-M90GoSM4nsig796nLRL:b-ajtd8) ([raw file](https://github.com/cockroachdb/cockroach/blob/52cf6aa18e52b5441da64b2430a99d1ec4c1ceff/pkg/sql/pgwire/types.go#L200)):*
> ```
> datum:oid
> ```
Should the comment say "datum:type mapping" now?
___
*[pkg/sql/pgwire/types.go, line 202 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/49886#-M90Hb_-5aXkctbT8yj-:-M90Hb_-5aXkctbT8yj0:b5keyiy) ([raw file](https://github.com/cockroachdb/cockroach/blob/52cf6aa18e52b5441da64b2430a99d1ec4c1ceff/pkg/sql/pgwire/types.go#L202)):*
> ```Go
> // (and can be nil) for types with a 1:1 datum:oid mapping.
> func (b *writeBuffer) writeBinaryDatum(
> 	ctx context.Context, d tree.Datum, sessionLoc *time.Location, T *types.T,
> ```

nit: parameter name should either be `t` or `type`.
___
*[pkg/sql/pgwire/types.go, line 412 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/49886#-M90GyEV8vX4NZaksg_b:-M90GyEV8vX4NZaksg_c:befo5u) ([raw file](https://github.com/cockroachdb/cockroach/blob/52cf6aa18e52b5441da64b2430a99d1ec4c1ceff/pkg/sql/pgwire/types.go#L412)):*
> ```Go
> 		//if T.Oid() == oid.T_bpchar {
> 		//	sv = fmt.Sprintf("%-*v", T.Width(), string(*v))
> 		//}
> ```

Looks like this was accidentally commented.

Which also raises the question -- is this line getting tested?
___
*[pkg/sql/pgwire/types_test.go, line 101 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/49886#-M90Hnq90Q76-jF20F-o:-M90Hnq90Q76-jF20F-p:b5139ys) ([raw file](https://github.com/cockroachdb/cockroach/blob/52cf6aa18e52b5441da64b2430a99d1ec4c1ceff/pkg/sql/pgwire/types_test.go#L101)):*
> ```Go
> 
> 	writeBuf1 := newWriteBuffer(nil /* bytecount */)
> 	writeBuf1.writeTextDatum(context.Background(), ary, defaultConv, nil /* T */)
> ```

nit: if you change the parameter name, then could you update the inline comments here and below?

**Review by arulajmani — COMMENTED** [review]
Thanks for the quick reviews! I'm having some trouble regenerating encodings.json when writing a test for the binary encoding bp_chars, but Ill update the PR with a test soon.

Addressed all the other comments though. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49886#-:-M90cJ8KAsnvQflx3svJ:bbey5xc)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @rafiss and @rohany)

___
*[pkg/sql/pgwire/command_result.go, line 81 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/49886#-M90FvEOW-o5sGZ52L0D-r1-81:-M90J3Rq3ukhDoZeWTy3:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/52cf6aa18e52b5441da64b2430a99d1ec4c1ceff/pkg/sql/pgwire/command_result.go#L81)):*
<details><summary><i>Previously, rohany (Rohan Yadav) wrote…</i></summary><blockquote>

You should use `*types.T` instead, theres no real reason to copy it.
</blockquote></details>

Done.
___
*[pkg/sql/pgwire/types.go, line 72 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/49886#-M90HU_t78nL-8OVI6CE:-M90KRne904y-UAv4w_I:b-6wggtu) ([raw file](https://github.com/cockroachdb/cockroach/blob/52cf6aa18e52b5441da64b2430a99d1ec4c1ceff/pkg/sql/pgwire/types.go#L72)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

> ```
> T
> ```
nit: the parameter name should be a lower-case `t`, though I'd even be more in favor of using the name `type`.
</blockquote></details>

Can't use `type` because its a reserved keyword, kinda unfortunate :/ Changing to `t`. 
___
*[pkg/sql/pgwire/types.go, line 124 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/49886#-M90FvEPwRtOFdjBNYSE-r1-124:-M90JxiLAuznW1H55MMr:b-5pr3in) ([raw file](https://github.com/cockroachdb/cockroach/blob/52cf6aa18e52b5441da64b2430a99d1ec4c1ceff/pkg/sql/pgwire/types.go#L124)):*
<details><summary><i>Previously, rohany (Rohan Yadav) wrote…</i></summary><blockquote>

Just checking, bpchar is the opposite of varchar right (using fixed width)? If so, this needs a comment, and so does the formatting directive here (idk what this does)
</blockquote></details>

Bpchar -- the bp in bp_char stands for blank padding, so Im abstracting the padding logic into `resolveBlankPaddedChar` and leaving a comment there. 
formatting -- it pads the string with spaces on the right, added a comment. 
___
*[pkg/sql/pgwire/types.go, line 200 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/49886#-M90GoSLBa0HauEPFn1_:-M90Jnh8C_HfJ4bMmoaJ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/52cf6aa18e52b5441da64b2430a99d1ec4c1ceff/pkg/sql/pgwire/types.go#L200)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

> ```
> datum:oid
> ```
Should the comment say "datum:type mapping" now?
</blockquote></details>

Done.
___
*[pkg/sql/pgwire/types.go, line 410 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/49886#-M90FvEPwRtOFdjBNYSE-r1-410:-M90MztuCMfw9UXxHcLR:b-fj9dgk) ([raw file](https://github.com/cockroachdb/cockroach/blob/52cf6aa18e52b5441da64b2430a99d1ec4c1ceff/pkg/sql/pgwire/types.go#L410)):*
<details><summary><i>Previously, rohany (Rohan Yadav) wrote…</i></summary><blockquote>

commented code?
</blockquote></details>

oops. Fixed and moved this to a function `resolveBlankPaddedChar`
___
*[pkg/sql/pgwire/types.go, line 412 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/49886#-M90GyEV8vX4NZaksg_b:-M90Sszx7rCrYqlvqP7U:b-edce4t) ([raw file](https://github.com/cockroachdb/cockroach/blob/52cf6aa18e52b5441da64b2430a99d1ec4c1ceff/pkg/sql/pgwire/types.go#L412)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

Looks like this was accidentally commented.

Which also raises the question -- is this line getting tested?
</blockquote></details>

Thats true. I'm getting an auth error when trying to regenerate encoding.json to add a test for this, but Ill resolve this and update the PR. 
___
*[pkg/sql/pgwire/testdata/pgtest/row_description, line 119 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/49886#-M90FvEPwRtOFdjBNYSD-r1-119:-M90NtIsBS9ieVbZkIx-:b-jgsiqh) ([raw file](https://github.com/cockroachdb/cockroach/blob/52cf6aa18e52b5441da64b2430a99d1ec4c1ceff/pkg/sql/pgwire/testdata/pgtest/row_description#L119)):*
<details><summary><i>Previously, rohany (Rohan Yadav) wrote…</i></summary><blockquote>

Are these meant to be changed? The comment for the bug is below at line 172
</blockquote></details>

Yes, the accidental comment in code was in `writeBinaryDatum`, but this test uses the `pgwirebase.FormatText` format code and therefore goes through `writeTextDatum` (where the padding change was present).

**Review by rafiss — COMMENTED** [review]
just a heads up -- you may want to rebase since the row_description test was recently refactored

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49886#-:-M9KXmZk9RUIJDLzUGoj:b2dfcx3)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @rafiss and @rohany)

**Comment by arulajmani** [comment]
Rebased + added new tests. Should be ready for review now!

File: pkg/sql/pgwire/pgwirebase/encoding.go [hunk 5]
```diff
@@ -671,11 +672,18 @@ func DecodeOidDatum(
...
 		if err := validateStringBytes(b); err != nil {
 			return nil, err
 		}
 		return tree.NewDString(string(b)), nil
+	case oid.T_bpchar:
+		if err := validateStringBytes(b); err != nil {
+			return nil, err
+		}
+		// Trim the trailing spaces
+		sv := strings.TrimRight(string(b), " ")
```
Comment (rohany) [inline pkg/sql/pgwire/pgwirebase/encoding.go:685]: Is there any way that empty spaces could be considered part of the input string?

File: pkg/sql/pgwire/testdata/pgtest/row_description [hunk 6]
```diff
@@ -171,16 +171,11 @@ RowDescription
 # The following discrepancy is a bug.
 # See: https://github.com/cockroachdb/cockroach/issues/49639
```
Comment (rohany) [inline pkg/sql/pgwire/testdata/pgtest/row_description:172]: Can delete this now right?

**Review by rohany — COMMENTED** [review]
It looks like you picked up some accidental changes in some proto files

**Review by arulajmani — COMMENTED** [review]
TFTR! Not sure how those proto file changes got picked up, but fixed. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49886#-:-M9OmOf9DoHcEbwKNhot:bc920iw)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @rafiss and @rohany)

___
*[pkg/sql/pgwire/pgwirebase/encoding.go, line 685 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/49886#-M9KogKXIdLW_jjbD9bR-r3-685:-M9Omgr80wRmu9kuVIwk:b-xbq740) ([raw file](https://github.com/cockroachdb/cockroach/blob/827ca3c2f73171307e6e0479e39e242c1f60d816/pkg/sql/pgwire/pgwirebase/encoding.go#L685)):*
I think empty spaces would just be considered trailing spaces. From https://www.postgresql.org/docs/9.1/datatype-character.html: 

> Values of type character are physically padded with spaces to the specified width n, and are stored and displayed that way. However, the padding spaces are treated as semantically insignificant. Trailing spaces are disregarded when comparing two values of type character, and they will be removed when converting a character value to one of the other string types.

So I think stripping away empty spaces should be fine. Not sure if there is a good way to test this though. 
___
*[pkg/sql/pgwire/testdata/pgtest/row_description, line 172 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/49886#-M90FvEPwRtOFdjBNYSD-r3-172:-M9OmYMG1ybRQYQvNcWi:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/827ca3c2f73171307e6e0479e39e242c1f60d816/pkg/sql/pgwire/testdata/pgtest/row_description#L172)):*
<details><summary><i>Previously, rohany (Rohan Yadav) wrote…</i></summary><blockquote>

Can delete this now right?
</blockquote></details>

Done.

**Comment by rohany** [comment]
> So I think stripping away empty spaces should be fine. Not sure if there is a good way to test this though.

Could you insert a `bpchar` with a bunch of spaces at the end?

**Review by arulajmani — COMMENTED** [review]
As discussed offline, it's hard to test `bpchars` with a bunch of spaces at the end because of the way we currently construct DString datums, which manifests itself in issues like #50015 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49886#-:-M9PGgPvA-01fHkl7op5:bzijen2)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @rafiss and @rohany)

**Review by solongordon — APPROVED** [review]
LGTM module a couple nits

> Reviewed 1 of 7 files at r1, 4 of 4 files at r2, 5 of 8 files at r3, 3 of 3 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49886#-:-M9TclBS4gT-SOYfZm25:b-joo1zm)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @arulajmani, @rafiss, and @rohany)

___
*[pkg/sql/pgwire/encoding_test.go, line 92 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/49886#-M9TbMffCA0b8pQnJskP:-M9TbMffCA0b8pQnJskQ:b9g6kfb) ([raw file](https://github.com/cockroachdb/cockroach/blob/0c065bfa49dc3054032a4ba747a1373da79643ee/pkg/sql/pgwire/encoding_test.go#L92)):*
> ```Go
> 		tc.T = types.OidToType[tc.Oid]
> 		if tc.T == nil {
> 			t.Fatalf("Unknown Oid %d not found in the OidToType map", tc.Oid)
> ```

Nit: CRDB convention is not to capitalize the first letter in errors
___
*[pkg/sql/pgwire/encoding_test.go, line 96 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/49886#-M9TbLpZBE86hfmuugPv:-M9TbLpZBE86hfmuugPw:b5jhvyq) ([raw file](https://github.com/cockroachdb/cockroach/blob/0c065bfa49dc3054032a4ba747a1373da79643ee/pkg/sql/pgwire/encoding_test.go#L96)):*
> ```Go
> 
> 		// Populate specific type information based on OID and the specific test
> 		//case.
> ```

Nit: missing space

**Comment by arulajmani** [comment]
TFTRs! Fixed nits, merging. 

bors r=solongordon

**Pull request merged**